### DISCUSSION TITLE
M1 Mac GitHub Action to Publish Website uses different version of Python than Quarto

### DISCUSSION THREAD

**Discussion by rressler** [quarto-dev/quarto-cli#3861] [Q&A] [answered]
When using a GitHub action per the reference to publish a website, it errors  because it is invoking a different version of Python that is on a different path and does not have jupyter it appears (quarto publish works fine). I tried using pyenv and other suggestions but cannot upgrade the version under the usr/bin/python3 that comes with the M1 Mac, nor add the nbformat to it. Note, when I actually run the usr/bin is says it is python 3.9.

Will an environment variable fix this? If so what should it look like to tell GitHub Actions the correct version to use?

Thanks,

Quarto Check shows:

```
Checking Python 3 installation....OK
      Version: 3.11.0
      Path: /Library/Frameworks/Python.framework/Versions/3.11/bin/python3
      Jupyter: 5.1.0
      Kernels: python3
```

GitHub Action to publish website shows 

```
Starting python3 kernel...Traceback (most recent call last):
  File "/opt/quarto/share/jupyter/jupyter.py", line 21, in <module>
    from notebook import notebook_execute, RestartKernel
  File "/opt/quarto/share/jupyter/notebook.py", line 16, in <module>
    import nbformat
ModuleNotFoundError: No module named 'nbformat'
Python 3 installation:
  Version: 3.10.6
  Path: /usr/bin/python3
  Jupyter: (None)

Jupyter is not available in this Python installation.
Install with python3 -m pip install jupyter
```

**Comment by cscheid** [comment]
A few notes/questions:

- Did you go over the documentation in https://github.com/quarto-dev/quarto-actions/, specifically the examples for the `setup` action? 
- Can you please share the repository whose actions are failing?

It seems like you're discussing a mix of locally installed software and errors on GitHub actions. A GitHub action executes on Github's servers, not on your computer, so changing your local installation will not change anything.

**Comment by rressler** [comment]
I did go over the documentation (I have not used other GitHub Actions) but it is a direct copy and past from the Quarto documentation for the action.

I did not set up a venv but perhaps I need to do that.
I sent an invite for collaboration on GitHub,

I can publish fine using quarto publish but it would be nice to get the GitHub action work to go to Quartopub.

Thanks,
Richard

From: Carlos Scheidegger ***@***.***>
Date: Thursday, January 5, 2023 at 12:05 PM
To: quarto-dev/quarto-cli ***@***.***>
Cc: Richard Ressler ***@***.***>, Author ***@***.***>
Subject: Re: [quarto-dev/quarto-cli] M1 Mac GitHub Action to Publish Website uses different version of Python than Quarto (Discussion #3861)

External Email: Use caution with links and attachments.

A few notes/questions:

  *   Did you go over the documentation in https://github.com/quarto-dev/quarto-actions/<https://urldefense.com/v3/__https:/github.com/quarto-dev/quarto-actions/__;!!IaT_gp1N!2vEk7FhPC4XNjvGh864vLeo7OEI8e_-mCT74HvgyNs7sbrqgNQgxCrYviWAMUCF3IDc5Q1zE_UaSgy5WhXxjTpQ7CCk$>, specifically the examples for the setup action?
  *   Can you please share the repository whose actions are failing?

It seems like you're discussing a mix of locally installed software and errors on GitHub actions. A GitHub action executes on Github's servers, not on your computer, so changing your local installation will not change anything.

**Reply by cscheid** [reply]
Take a look at https://github.com/quarto-dev/quarto-actions/tree/main/examples, specifically the "dependencies" part. If you directly copy-pasted only the action in setup you are likely missing the steps needed to install Python.

**Comment by rressler** [comment, marked as answer]
Carlos,
Thanks, I set up the full venv with requirements.txt and then tried just with the dependencies code. I was getting an error on not finding the pip cache so I changed to runs-on: macos-latest
and the issues with pip cache went away so all works.

I was not expecting any issues with python since I don’t have any python code in the website but I guess there is something in the render process that requires it

Thanks for the assist,
Richard



From: Carlos Scheidegger ***@***.***>
Date: Friday, January 6, 2023 at 10:29 AM
To: quarto-dev/quarto-cli ***@***.***>
Cc: Richard Ressler ***@***.***>, Author ***@***.***>
Subject: Re: [quarto-dev/quarto-cli] M1 Mac GitHub Action to Publish Website uses different version of Python than Quarto (Discussion #3861)

External Email: Use caution with links and attachments.

Take a look at https://github.com/quarto-dev/quarto-actions/tree/main/examples<https://urldefense.com/v3/__https:/github.com/quarto-dev/quarto-actions/tree/main/examples__;!!IaT_gp1N!3HCBQiDJ83tYxms2ZfvpJH0w-QuJMfJ2d3NSFsYGmlyE-S8kFHzJvwyjfUkw52qMp_vvOfkp9969NpzERrIP32pHm34$>, specifically the "dependencies" part. If you directly copy-pasted only the action in setup you are likely missing the steps needed to install Python.

**Reply by cscheid** [reply]
You do seem to be using the jupyter engine, which does require you to install some python dependencies, namely `nbformat`. Otherwise, you wouldn't see this

```
Starting python3 kernel...Traceback (most recent call last):
  File "/opt/quarto/share/jupyter/jupyter.py", line 21, in <module>
    from notebook import notebook_execute, RestartKernel
  File "/opt/quarto/share/jupyter/notebook.py", line 16, in <module>
    import nbformat
ModuleNotFoundError: No module named 'nbformat'
```
